src/GitHub.Exports.Reactive/Collections/TrackingCollection.cs

```diff
@@ -0,0 +1,842 @@
+            }
+        }
+
+        void SetAndRecalculateFilter(Func<T, int, IList<T>, bool> newFilter)
+        {
+            if (filter == null && newFilter == null)
+                return; // nothing to do
+
+            ClearItems();
+            filter = newFilter;
+            RecalculateFilter(original, 0, 0, original.Count);
+        }
+
+        protected override void ClearItems()
+        {
+            filteredIndexCache.Clear();
+            base.ClearItems();
+        }
+
+        ActionData CheckFilter(ActionData data)
+        {
+            data.IsIncluded = true;
+            if (filter != null)
+                data.IsIncluded = filter(data.Item, data.Position, this);
+            return data;
+        }
+
+        int StartQueue()
+        {
+            disposables.Add(sourceQueue.Subscribe(_ => itemCount++));
+            return 0;
+        }
+
+        T GetFromQueue()
+        {
+            try
+            {
+                T d = null;
+                if (queue?.TryDequeue(out d) ?? false)
+                    return d;
+            }
+            catch { }
+            return null;
+        }
+
+        ActionData ProcessItem(T item, List<T> list)
+        {
+            ActionData ret;
+
+            var idx = GetIndexUnfiltered(item);
+            if (idx >= 0)
+            {
+                var old = list[idx];
+                var comparison = comparer(item, old);
+
+                // no sorting to be done, just replacing the element in-place
+                if (comparison == 0)
+                    ret = new ActionData(TheAction.None, item, null, idx, idx, list);
+                else
+                    // element has moved, save the original object, because we want to update its contents and move it
+                    // but not overwrite the instance.
+                    ret = new ActionData(TheAction.Move, item, old, comparison, idx, list);
+            }
+            // the element doesn't exist yet
+            // figure out whether we're larger than the last element or smaller than the first or
+            // if we have to place the new item somewhere in the middle
+            else if (list.Count > 0)
+            {
+                if (comparer(list[0], item) >= 0)
+                    ret = new ActionData(TheAction.Insert, item, null, 0, -1, list);
+
+                else if (comparer(list[list.Count - 1], item) <= 0)
+                    ret = new ActionData(TheAction.Add, item, null, list.Count, -1, list);
+
+                // this happens if the original observable is not sorted, or it's sorting order doesn't
+                // match the comparer that has been set
+                else
+                {
+                    idx = BinarySearch(list, item, comparer);
+                    if (idx < 0)
+                        idx = ~idx;
+                    if (idx == list.Count)
+                        ret = new ActionData(TheAction.Add, item, null, list.Count, -1, list);
```

Comment:
This code path is never touched by the unit tests. We should make sure a test exercises it.